src/main/scala/com/microsoft/ml/spark/explainers/KernelSHAPBase.scala

```diff
@@ -0,0 +1,150 @@
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+trait KernelSHAPParams extends HasNumSamples with HasMetricsCol {
+  self: KernelSHAPBase =>
+
+  val infWeight = new DoubleParam(
+    this,
+    "infWeight",
+    "The double value to represent infinite weight. Default: 1E8.",
+    ParamValidators.gtEq(1)
+  )
+
+  def getInfWeight: Double = $(infWeight)
+
+  def setInfWeight(value: Double): this.type = this.set(infWeight, value)
+
+  setDefault(metricsCol -> "r2", infWeight -> 1E8)
+}
+
+abstract class KernelSHAPBase(override val uid: String)
+  extends LocalExplainer
+    with KernelSHAPParams
+    with Wrappable
+    with BasicLogging {
+
+  override def transform(instances: Dataset[_]): DataFrame = logTransform {
+    import instances.sparkSession.implicits._
+    this.validateSchema(instances.schema)
+
+    val df = instances.toDF
+    val idCol = DatasetExtensions.findUnusedColumnName("id", df)
+    val coalitionCol = DatasetExtensions.findUnusedColumnName("coalition", df)
+    val weightCol = DatasetExtensions.findUnusedColumnName("weight", df)
+
+    val dfWithId = df.withColumn(idCol, monotonically_increasing_id())
+    val preprocessed = preprocess(dfWithId).cache()
+
+    val sampleWeightUdf = UDFUtils.oldUdf(kernelWeight(this.getInfWeight) _, DoubleType)
+    val samples = createSamples(preprocessed, idCol, coalitionCol)
+      .repartition()
+
+    val scored = getModel.transform(samples)
+    val explainTargetCol = DatasetExtensions.findUnusedColumnName("target", scored)
+
+    val coalitionScores = scored
+      .withColumn(explainTargetCol, this.getExplainTarget(scored.schema))
+      .groupBy(col(idCol), col(coalitionCol))
+      .agg(Summarizer.mean(col(explainTargetCol)).alias(explainTargetCol))
+      .withColumn(weightCol, sampleWeightUdf(col(coalitionCol)))
+
+    val fitted = coalitionScores.groupByKey(row => row.getAs[Long](idCol)).mapGroups {
+      case (id: Long, rows: Iterator[Row]) =>
+        val (inputs, outputs, weights) = rows.map {
+          row =>
+            val input = row.getAs[SV](coalitionCol).toBreeze
+            val output = row.getAs[SV](explainTargetCol).toBreeze
+            val weight = row.getAs[Double](weightCol)
+            (input, output, weight)
+        }.toSeq.unzip3
+
+        val inputsBV = BDM(inputs: _*)
+        val outputsBV = BDM(outputs: _*)
+        val weightsBV = BDV(weights: _*)
+
+        val wlsResults = outputsBV(::, *).toIndexedSeq.map {
+          new LeastSquaresRegression().fit(inputsBV, _, weightsBV, fitIntercept = true)
+        }
+
+        val coefficientsMatrix = wlsResults.map(r => SVS.dense(r.intercept, r.coefficients.toArray: _*))
+        val metrics = BDV(wlsResults.map(_.rSquared): _*)
+        (id, coefficientsMatrix, metrics.toSpark)
+    }.toDF(idCol, this.getOutputCol, this.getMetricsCol)
+
+    preprocessed.hint("broadcast").join(fitted, Seq(idCol), "inner").drop(idCol)
```

Comment:
is this the same as calling:
```
import org.apache.spark.sql.functions.broadcast
broadcast(preprocessed).join(fitted, Seq(idCol), "inner").drop(idCol)
```
https://spark.apache.org/docs/2.4.0/sql-performance-tuning.html

Reply:
Spark 3 seems to recommend `df.hint("broadcast")` syntax:
https://spark.apache.org/docs/3.1.1/sql-performance-tuning.html#join-strategy-hints-for-sql-queries

`df.hint("broadcast")` creates an UnresolvedHint logical plan, while `broadcast(df)` creates a ResolvedHint logical plan. I cannot find any document regarding the difference between the two, so I'm sticking with what Spark 3 recommends.